rock-n-rollup/src/core/runtime.rs

```diff
@@ -104,132 +43,126 @@ pub trait Runtime: 'static {
+            Ok(_) => Ok(()),
+            Err(_) => {
+                self.host.write_debug("Error store_delete");
+                Err(())
+            }
         }
     }
 
     fn store_read(&mut self, path: &str, offset: usize, size: usize) -> Option<Vec<u8>> {
-        if !self.store_is_present(path) {
-            return None;
-        }
+        let path = OwnedPath::try_from(path.to_string()).map_err(|_| ()).ok()?;
 
-        let ptr = path.as_ptr();
-        let path_len = path.len();
-        let mut buffer = Vec::with_capacity(size);
-        let dst = buffer.as_mut_ptr();
-        unsafe {
-            let _ = store_read(ptr, path_len, offset, dst, size);
-            buffer.set_len(size);
-        }
-
-        Some(buffer)
+        self.host.store_read(&path, offset, size).ok()
     }
 
     fn store_write(&mut self, path: &str, data: &[u8], at_offset: usize) -> Result<(), ()> {
-        let res = unsafe {
-            let path_len = path.len();
-            let path = path.as_ptr();
-            let num_bytes = data.len();
-            let src = data.as_ptr();
-            store_write(path, path_len, at_offset, src, num_bytes)
-        };
+        let path = OwnedPath::try_from(path.to_string()).map_err(|_| ())?;
+
+        let res = self.host.store_write(&path, data, at_offset);
         match res {
-            0 => Ok(()),
-            err => {
-                self.write_debug(&format!("error store_write_raw: {}\n", err));
+            Ok(_) => Ok(()),
+            Err(_) => {
+                self.host.write_debug("Error store_write");
                 Err(())
             }
         }
     }
-    fn reveal_preimage(&mut self, hash: &[u8; PREIMAGE_HASH_SIZE]) -> Result<Vec<u8>, ()> {
-        let max_size = 4096;
-        let mut payload = Vec::with_capacity(MAX_MESSAGE_SIZE as usize);
 
-        let u8_size = u8::try_from(PREIMAGE_HASH_SIZE).unwrap();
+    fn store_move(&mut self, from: &str, to: &str) -> Result<(), ()> {
+        let from = OwnedPath::try_from(from.to_string()).map_err(|_| ())?;
+        let to = OwnedPath::try_from(to.to_string()).map_err(|_| ())?;
 
-        unsafe {
-            let size = reveal_preimage(hash.as_ptr(), u8_size, payload.as_mut_ptr(), max_size);
-            if size < 0 {
+        let res = self.host.store_move(&from, &to);
+        match res {
+            Ok(_) => Ok(()),
+            Err(_) => {
+                self.host.write_debug("Error store_move");
                 Err(())
-            } else {
-                let size = usize::try_from(size).unwrap();
-                payload.set_len(size);
-                Ok(payload)
             }
         }
     }
 
-    fn store_move(&mut self, from: &str, to: &str) -> Result<(), ()> {
-        let res = unsafe { store_move(from.as_ptr(), from.len(), to.as_ptr(), to.len()) };
+    fn reveal_preimage(&mut self, hash: &[u8; PREIMAGE_HASH_SIZE]) -> Result<Vec<u8>, ()> {
+        let payload: Vec<u8> = Vec::with_capacity(MAX_MESSAGE_SIZE);
```

Comment:
I am wondering if the vector can be initialized with Vec::new(), or not. But because I didn't add tests it's difficult to say